### PR TITLE
Add requestDate parameter to presignedUrl() and presignedGetObject()

### PR DESCRIPTION
**Pull request by dustin-H**
Hi there,

I just added a parameter `requestDate` to the methods `presignedUrl()` and `presignedGetObject()`.

Please have a look at #724 for more details.

My todo checklist:

- [x] Adjust the code
- [x] Add js-doc parameter description
- [x] Create new tests for the functionality
- [x] All (old and new) tests should pass
- [x] Update docs
- [x] Create example

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -702,7 +702,7 @@ minioClient.removeIncompleteUpload('mybucket', 'photo.jpg', function(err) {
 Presigned URLs are generated for temporary download/upload access to private objects.
 
 <a name="presignedUrl"></a>
-### presignedUrl(httpMethod, bucketName, objectName, expiry[, reqParams, cb])
+### presignedUrl(httpMethod, bucketName, objectName, expiry[, reqParams, requestDate, cb])
```
Comment (ebozduman) [inline docs/API.md]: `expiry` as also an optional parameter. `[` is supposed to include it too.

File: docs/API.md [hunk 2]
```diff
@@ -761,6 +762,8 @@ __Parameters__
 |`bucketName` | _string_ | Name of the bucket. |
 |`objectName` | _string_ | Name of the object. |
 |`expiry`     | _number_ | Expiry time in seconds. Default value is 7 days. |
+|`respHeaders`  | _object_ | response headers to override (optional) |
+|`requestDate`  | _Date_ | A date object, the url will be issued at. Default value is now. |
```
Comment (ebozduman) [inline docs/API.md]: Title above also needs to be modified accordingly, line# 751:

`### presignedGetObject(bucketName, objectName, expiry[, cb])`
=>
`### presignedGetObject(bucketName, objectName[, expiry, respHeaders, requestDate, cb])`

`(optinal)` information should be added to the above `expiry` and `requestDate` descriptions.

File: docs/API.md [hunk 3]
```diff
@@ -717,6 +717,7 @@ __Parameters__
 |`objectName` | _string_ | Name of the object. |
 |`expiry`     | _number_ | Expiry time in seconds. Default value is 7 days. |
 |`reqParams`  | _object_ | request parameters. |
+|`requestDate`  | _Date_ | A date object, the url will be issued at. Default value is now. |
```
Comment (ebozduman) [inline docs/API.md]: `(optinal)` information should be added to the above `expiry`, `reqParams`, and `requestDate` descriptions.

**Comment by dustin-H** [comment]
@ebozduman 
Yes, I've seen this.

However, I think this is a yarn issue https://github.com/yarnpkg/yarn/issues/6619.

I haven't changed anything regarding this.

**Comment by dustin-H** [comment]
@ebozduman 
I fixed it by setting the yarn version in `.travis.yml` to `yarn@1.9.4`.

This is the latest version, which has successfully passed the last travis build.

**Comment by ebozduman** [comment]
@dustin-H,
Perfect. Great job!

File: src/main/minio.js [hunk 4]
```diff
@@ -1602,7 +1612,7 @@ export class Client {
         throw new TypeError(`response header ${header} should be of type "string"`)
       }
     })
-    return this.presignedUrl('GET', bucketName, objectName, expires, respHeaders, cb)
+    return this.presignedUrl('GET', bucketName, objectName, expires, respHeaders, requestDate, cb)
```
Comment (ebozduman) [inline src/main/minio.js]: @dustin-H,
When you filed the issue for this PR, #724, as expected, you've also asked for `requestDate` parameter to be added for `presignedPutObject` api, but your fix does not include it.

Comment (ebozduman) [reply src/main/minio.js]: @dustin-H,
Makes sense.
I'll be testing your change today (which is required before giving a LGTM) and then give it a go.

Comment (ebozduman) [reply src/main/minio.js]: @dustin-H,
Makes sense.

Today, I'll be testing the fix (which is supposed to be done before giving a LGTM), and then give it a go.

File: examples/presigned-getobject-request-date.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+var s3Client = new Minio.Client({
+  endPoint: 's3.amazonaws.com',
+  accessKey: 'YOUR-ACCESSKEYID',
+  secretKey: 'YOUR-SECRETACCESSKEY',
+  useSSL: true // Default is true.
+})
+
+// Presigned get object URL for my-objectname at my-bucketname, it expires in 7 days by default.
+var requestDate = new Date('2017-01-01');
+var presignedUrl = s3Client.presignedGetObject('my-bucketname', 'my-objectname', 1000, requestDate, function(e, presignedUrl) {
```
Comment (ebozduman) [inline examples/presigned-getobject-request-date.js]: I've tested the fix using this example script. However, it kept failing.
The reason turned out to be the missing optional parameter `respHeaders` in the call. So, `requestDate` value was read as `respHeaders` and `requestDate` got the default `now` value. 
We have to add the `respHeaders` in the call:
`s3Client.presignedGetObject('my-bucketname', 'my-objectname', 1000, requestDate, ...`
=>
`s3Client.presignedGetObject('my-bucketname', 'my-objectname', 1000,`**`{},`**`requestDate, ...`

Comment (dustin-H) [reply examples/presigned-getobject-request-date.js]: Sorry for that.

Fixed in 02331c94739b0e563a35e0c35bb82d615c1d2ab7

File: src/test/functional/functional-tests.js [hunk 6]
```diff
@@ -751,6 +751,31 @@ describe('functional tests', function() {
       })
     })
 
+    step(`presignedUrl(httpMethod, bucketName, objectName, expires, cb)_httpMethod:GET, bucketName:${bucketName}, objectName:${_1byteObjectName}, expires:86400, requestDate:StartOfDay_`, done => {
```
Comment (krishnasrinivas) [inline src/test/functional/functional-tests.js]: presignedUrl() is an undocumented internal function. Hence this functional test need can be deleted.

Comment (dustin-H) [reply src/test/functional/functional-tests.js]: Hi @krishnasrinivas,

This function is public documented here: https://docs.minio.io/docs/javascript-client-api-reference.html#presignedUrl

**Comment by dustin-H** [comment]
@ebozduman @krishnasrinivas 

Do you know, when this will be merged. I really need this.

**Comment by kannappanr** [comment]
@dustin-H Thanks for the PR. This has been merged. We will make a release in the next few days.

**Pull request merged**